File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/loader/ProxyClassLoader.java
package net.yakclient.mixin.base.internal.loader;

import java.nio.ByteBuffer;
import java.security.CodeSource;
import java.security.ProtectionDomain;
import java.security.cert.Certificate;

public class ProxyClassLoader extends ClassLoader {
    private static final String CONTEXT_POOL_MANAGER = "net.yakclient.mixin.internal.loader.ContextPoolManager";
    protected final ProtectionDomain defaultDomain =
            new ProtectionDomain(new CodeSource(null, (Certificate[]) null),
                    null, this, null);

    public ProxyClassLoader(ClassLoader parent) {
        super(parent);
        Thread.currentThread().setContextClassLoader(this);
    }

    public Class<?> defineClass(String name, byte[] b) {
        return this.defineClass(name, ByteBuffer.wrap(b), this.defaultDomain);
    }
//
//
    public boolean isDefined(String cls) {
        return super.findLoadedClass(cls) != null;
    }

    @Override
    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
        if (this.isDefined(CONTEXT_POOL_MANAGER) && ContextPoolManager.isTargeted(name))
            return ContextPoolManager.loadClass(name);
        return super.loadClass(name, resolve);
    }

}

File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/internal/loader/ProxyClassLoader.kt
package net.yakclient.mixin.base.internal.loader

interface ProxyClassLoader {
//    protected val defaultDomain = ProtectionDomain(
//        CodeSource(null, null as Array<Certificate?>?),
//        null, this, null
//    )

    fun defineClass(name: String, b: ByteArray): Class<*>
//
//    {
//        return super.defineClass(name, b, 0, b.size)
//    }

    //
    //
    fun isDefined(cls: String): Boolean

    fun loadClass(name: String): Class<*>

//    {
//        return super.findLoadedClass(cls) != null
//    } //    @Override

    //    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
    //        if (this.isDefined(CONTEXT_POOL_MANAGER) && ContextPoolManager.isTargeted(name))
    //            return ContextPoolManager.loadClass(name);
    //        return super.loadClass(name, resolve);
    //    }

}